examples/shm_throughput/ShmThroughput.idl

```diff
@@ -1,146 +1,145 @@
 module ThroughputModule
 {
-
-struct DataType_Base
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-};
-#pragma keylist DataType_Base
-
-struct DataType_16
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[16 - 12];
-};
-#pragma keylist DataType_16
-
-struct DataType_32
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[32 - 12];
-};
-#pragma keylist DataType_32
-
-struct DataType_64
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[64 - 12];
-};
-#pragma keylist DataType_64
-
-struct DataType_128
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[128 - 12];
-};
-#pragma keylist DataType_128
-
-struct DataType_256
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[256 - 12];
-};
-#pragma keylist DataType_256
-
-struct DataType_512
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[512 - 12];
-};
-#pragma keylist DataType_512
-
-struct DataType_1024
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[1024 - 12];
-};
-#pragma keylist DataType_1024
-
-struct DataType_2048
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[2048 - 12];
-};
-#pragma keylist DataType_2048
-
-struct DataType_4096
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[4096 - 12];
-};
-#pragma keylist DataType_4096
-
-struct DataType_8192
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[8192 - 12];
-};
-#pragma keylist DataType_8192
-
-struct DataType_16384
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[16384 - 12];
-};
-#pragma keylist DataType_16384
-
-struct DataType_32768
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[32768 - 12];
-};
-#pragma keylist DataType_32768
-
-struct DataType_65536
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[65536 - 12];
-};
-#pragma keylist DataType_65536
-
-struct DataType_131072
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[131072 - 12];
-};
-#pragma keylist DataType_131072
-
-struct DataType_262144
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[262144 - 12];
-};
-#pragma keylist DataType_262144
-
-struct DataType_524288
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[524288 - 12];
-};
-#pragma keylist DataType_524288
-
-struct DataType_1048576
-{
-  unsigned long long count;
-  unsigned long payloadsize;
-  octet payload[1048576 - 12];
-};
-#pragma keylist DataType_1048576
+  @final
```

Comment:
No key fields here or #pragma keylist, so shouldn't there be an @topic before each type which we want to make into a topic?

Reply:
By default, types are considered topic types (the default of the `@default_nested` annotation is `false`), so there is no need to explicitly add `@topic`.

Reply:
It just doesn't feel right that 'by default' all types (that are not explicitly stated as being 'nested') are considered to be/become topics i.e. require descriptor- and code-generation. At least that wasn't the case in the pre-Xtypes era.
Apparently its also stated that an IDL-compiler 'shall' have the ability to 'overrule' that default_nested behavior at 'top-level', but that sounds like 'self-inflicted damage'. 
So summarizing, I think it would really makes sense (to avoid unnecessary descriptor/code-generation that might be 'huge' given some elaborate data-models) to have to explicitly specify which types are meant to be/become topics.

And in-case that's not achievable, perhaps 'our' IDL-compiler could 'by-default' overrule the default_nested definition to achieve the same/desirable behavior.

Also curious about your opinion @eboasson ...

Reply:
I think the desired behaviour is mostly covered by:

> 7.3.1.2.1.16 Explicit declaration of Topic types
> A type may be explicitly annotated to indicate the intent to use it as a top-level DDS Topic. Absent this annotation only non-nested types may be used as DDS Topics. The@topic annotation overrides that default behavior.

because a carefully structured set of IDL files will usually be fine under this rule. However if I'd have a library of common type definitions and then (say) one IDL file per topic each using a subset of the types in that common library, it'd still produce a mess. If I am right in that, I wouldn't mind a mode that says "if there are types annotated with @topic, only those types may be used as DDS Topics".

Reply:
indeed, but that i.m.o. sensible mode would imply that that library should be _explicitly_ stated to include only nested types as by default each type would be treated as a non-nested type i.e. a to-become-topic (and thus producesthat mess)

Reply:
I suppose that's what `@default_nested` is for ... and then you end up with the current behaviour being good enough.

Reply:
thats  fine .. just noticing that more people are placing question-marks with sensible default behavior, see https://issues.omg.org/issues/DDSXTY14-15

Reply:
According to that issue, the spec already says

> In addition to the above annotations, IDL compilers shall provide the means to change the default value for non-annotated top-level modules

which seems to allow you to do what you are looking for. I don't think IDLC supports that yet, but a possible implementation could be to have a command-line option for controlling the default of `@default_nested` very much in the same vein as what this PR does for the default extensibility.

Reply:
Today I also read the `@nested` and `@topic` part in the IDL specification and it still doesn't make it clear how exactly the two interact. Listing all topic types (to me) seems more logical than listing all non-topic (`@nested`) types, but how that is controlled (idlc switch `--honor-topic / --honor-nested` possibly, but that seems weird).